lib/Wikia/src/Util/SamplerProxy.php

```diff
@@ -0,0 +1,134 @@
+<?php
+/**
+ * A wrapper for a POPO that supports dispatching a specific method to an alternate
+ * object, either as a shadowed execution (i.e. execute both the original method and the
+ * alternate) or as a substitution (i.e. execute the alternate instead of the original).
+ * Both shadowing and substitution may be sampled so that only a specified percentage of
+ * calls involve the alternate and the rest pass through to the original.
+ */
+namespace Wikia\Util;
+
+
+class SamplerProxy {
+
+	/**
+	 * @var string
+	 */
+	protected $enableShadowing;
+
+	/**
+	 * @var string
+	 */
+	protected $methodSamplingRate;
+
+	/**
+	 * @var callable
+	 * Expected to be of the form [ <instance>, <method name> ]
+	 */
+	protected $originalCallable;
+
+	/**
+	 * @var callable
+	 * Expected to be of the form [ <instance>, <method name> ]
+	 * Should have the same argument list as $originalCallable
+	 */
+	protected $alternateCallable;
+
+	/**
+	 * @var callable
+	 * Takes 2 arguments, ( $originalResults, $alternateResults ) and
+	 * returns the results that should be passed back to the caller
+	 */
+	protected $resultsCallable;
+
+	public static function createBuilder() {
+		return new SamplerProxyBuilder();
+	}
+
+	public function __construct( SamplerProxyBuilder $samplerProxyBuilder ) {
+		$this->enableShadowing = $samplerProxyBuilder->getEnableShadowing();
+		$this->methodSamplingRate = $samplerProxyBuilder->getMethodSamplingRate();
+		$this->originalCallable = $samplerProxyBuilder->getOriginalCallable();
+		$this->alternateCallable = $samplerProxyBuilder->getAlternateCallable();
+		$this->resultsCallable = $samplerProxyBuilder->getResultsCallable();
+	}
+
+	public function __call( $method, $args ) {
```

Comment:
I like general approach for this, however, this call got overly complex. Mixing sampling, dispatching and shadowing make this method very hard to understand and figure out, what in fact will happen when used. Additionally, I'd suggest not to include `resultsCallable` in here, it causes additional issues without providing actual value in terms of simplicity, code length or readability.

If you'd like to keep this utility as a piece for others to use, it should be simpler and straightforward. Make sure you ask for code review someone from platform team as well.